Easydict/Feature/ViewController/Window/WindowManager/EZWindowManager.m

```diff
@@ -912,31 +911,42 @@ - (void)closeFloatingWindow {
     
     // stop playing audio
     [self.floatingWindow.queryViewController stopPlayingQueryText];
-    
     self.floatingWindow.titleBar.pin = NO;
+    
+    EZBaseQueryWindow *closingFloatingWindow = self.floatingWindow;
+    
+    /// !!!: Close window may call window delegate method `windowDidResignKey:`
+    /// `windowDidResignKey:` will call `closeFloatingWindowIfNotPinnedOrMain`
     [self.floatingWindow close];
     
     if (![self currentShowingSettingsWindow]) {
         // recover last app.
         [self activeLastFrontmostApplication];
     }
     
-    if ([EZMainQueryWindow isAlive]) {
-        [self.mainWindow orderBack:nil];
+    /// !!!: Maybe floating window has closed after calling [self.floatingWindow close], so we need to check it.
+    if (self.floatingWindow == closingFloatingWindow) {
+        [self updateFloatingWindowType:self.floatingWindowType isShowing:NO];
     }
-    
-    // Move floating window type to second.
-    
-    NSNumber *windowType = @(self.floatingWindowType);
-    [self.floatingWindowTypeArray removeObject:windowType];
-    [self.floatingWindowTypeArray insertObject:windowType atIndex:1];
 }
 
-/// Close floating window, except main window.
-- (void)closeFloatingWindowExceptMain {
-    // Do not close main window
-    if (!self.floatingWindow.pin && self.floatingWindow.windowType != EZWindowTypeMain) {
-        [[EZWindowManager shared] closeFloatingWindow];
+/**
+ Close floating window if not pinned or main window.
+ Main window is basically equivalent to a pinned floating window.
+ */
+- (void)closeFloatingWindowIfNotPinnedOrMain {
+    [self closeFloatingWindowIfNotPinnedAndExceptWindowType:EZWindowTypeMain];
+}
+
+- (void)closeFloatingWindowIfNotPinned {
+    [self closeFloatingWindowIfNotPinnedAndExceptWindowType:EZWindowTypeNone];
+}
+
+- (void)closeFloatingWindowIfNotPinnedAndExceptWindowType:(EZWindowType)windowType {
+    BOOL isPinned = self.floatingWindow.isPin;
+    BOOL isTypeMatch = self.floatingWindow.windowType == windowType;
+    if (!isPinned && !isTypeMatch) {
```

Comment:
I wonder if we need this `windowType` param here since we generally exclude the main type only for now.

Reply:
Yes, but sometimes we don't need to exclude the main window, it's here to reduce duplicate code.

```objc
- (void)closeFloatingWindowIfNotPinnedOrMain {
    [self closeFloatingWindowIfNotPinnedAndExceptWindowType:EZWindowTypeMain];
}

- (void)closeFloatingWindowIfNotPinned {
    [self closeFloatingWindowIfNotPinnedAndExceptWindowType:EZWindowTypeNone];
}

- (void)closeFloatingWindowIfNotPinnedAndExceptWindowType:(EZWindowType)windowType {
    BOOL isPinned = self.floatingWindow.isPin;
    BOOL isTypeMatch = self.floatingWindow.windowType == windowType;
    if (!isPinned && !isTypeMatch) {
        [self closeFloatingWindow];
    }
}
```

Reply:
ok, I get it. 

Could we move `Not Pinned` check to `closeFloatingWindow`, it looks like we should never close a pinned window in any case.

Reply:
No, in some cases, even if it is a pinned window, we still need to be able to close it. 

For example, manually using the escape key or using shortcut keys to show/close the query window. You can refer to the invocation of the `closeFloatingWindow` method for details.

These are existing interaction logics that indicate users indeed want to manually close this window. I think we should not disrupt it.

Reply:
Update some code, let me know what you think.

Reply:
ok